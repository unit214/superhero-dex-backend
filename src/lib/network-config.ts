type NetworkConfig = {
  nodeUrl: string;
  compilerUrl: string;
  middlewareWebsocketUrl: string;
  middlewareHttpUrl: string;
};

export default {
  testnet: {
    nodeUrl: 'https://testnet.aeternity.io',
    compilerUrl: 'https://latest.compiler.aepps.com',
    middlewareWebsocketUrl: 'wss://testnet.aeternity.io/mdw/v2/websocket',
    middlewareHttpUrl: 'https://testnet.aeternity.io/mdw',
  },
  mainnet: {
    nodeUrl: 'https://mainnet.aeternity.io',
    compilerUrl: 'https://latest.compiler.aepps.com',
    middlewareWebsocketUrl: 'wss://mainnet.aeternity.io/mdw/v2/websocket',
    middlewareHttpUrl: 'https://mainnet.aeternity.io/mdw',
  },
} as { [key: string]: NetworkConfig };
