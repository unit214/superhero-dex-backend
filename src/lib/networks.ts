type NetworkConfig = {
  nodeUrl: string;
  compilerUrl: string;
  middlewareUrl: string;
  middlewareHttpUrl: string;
};

export default {
  testnet: {
    nodeUrl: 'https://testnet.aeternity.io',
    compilerUrl: 'https://latest.compiler.aepps.com',
    middlewareUrl: 'wss://testnet.aeternity.io/mdw/v2/websocket',
    middlewareHttpUrl: 'https://testnet.aeternity.io/mdw',
  },
  mainnet: {
    nodeUrl: 'https://mainnet.aeternity.io',
    compilerUrl: 'https://latest.compiler.aepps.com',
    middlewareUrl: 'wss://mainnet.aeternity.io/mdw/v2/websocket',
    middlewareHttpUrl: 'https://mainnet.aeternity.io/mdw',
  },
} as { [key: string]: NetworkConfig };
