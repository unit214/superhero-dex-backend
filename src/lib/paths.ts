export type Edge<E, N> = {
  data: E;
  t0: N;
  t1: N;
};

const includes =
  <N>(y: N) =>
  (t0: N, t1: N) => {
    if (t0 === y) {
      return -1;
    }
    if (t1 === y) {
      return 1;
    }
    return 0;
  };

type Graph<N extends string | number | symbol, E> = Record<
  N,
  Record<N, E | undefined> | undefined
>;

const getGraph = <E, N extends string | number | symbol>(
  start: N,
  edges: Edge<E, N>[],
): [Graph<N, E>, [N, E][]] => {
  const graph: Graph<N, E> = {} as any;
  const startEdges: [N, E][] = [];
  const addEdge = (n0: N, n1: N, data: E) => {
    if (!graph[n0]) {
      graph[n0] = {} as any;
    }
    (graph[n0] as any)[n1] = data;
  };
  for (const { t0, t1, data } of edges) {
    const hasStart = includes(start)(t0, t1);
    if (hasStart) {
      startEdges.push([hasStart === -1 ? t1 : t0, data]);
      continue;
    }
    addEdge(t0, t1, data);
    addEdge(t1, t0, data);
  }

  return [graph, startEdges];
};

export const getPaths = <E, N extends string | number | symbol>(
  start: N,
  end: N,
  edges: Edge<E, N>[],
): E[][] => {
  const [graph, startEdges] = getGraph(start, edges);
  return startEdges.reduce((acc: E[][], [middleNode, data]) => {
    //this is the direct connection
    if (middleNode === end) {
      return [[data]].concat(acc);
    }
    const middleEdge = graph[middleNode];
    const endData: E | undefined = middleEdge?.[end];
    return endData == null ? acc : acc.concat([[data, endData]]);
  }, []);
};
