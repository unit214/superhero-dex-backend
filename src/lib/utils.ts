import { Decimal } from '@prisma/client/runtime/library';
import BigNumber from 'bignumber.js';

export const nonNullable = <T>(t: T | null | undefined, label?: string): T => {
  if (t == null) {
    throw new Error(
      `${label ? label + ': ' : null} non nullable value expected`,
    );
  }
  return t;
};

export const removeId = <ID, T extends { id: ID }>(t: T) => {
  const {
    id, // eslint-disable-line @typescript-eslint/no-unused-vars
    ...tail
  } = t;
  return tail;
};

export const pluralize = (count: number, noun: string, suffix = 's') =>
  `${count} ${noun}${count !== 1 ? suffix : ''}`;

const parseEnv = (x) => x && JSON.parse(x);
export const presentInvalidTokens = parseEnv(process.env.SHOW_INVALID_TOKENS);

export const numberToDecimal = (number: number): Decimal =>
  new Decimal(number.toString());

export const bigIntToDecimal = (bigInt: bigint): Decimal =>
  new Decimal(bigInt.toString());

export const decimalToBigInt = (decimal: Decimal): bigint =>
  BigInt(decimal.toFixed().toString());

export const calculateUsdValue = ({
  reserve,
  tokenAePrice,
  decimals,
  aeUsdPrice,
}: {
  reserve: string;
  tokenAePrice: string;
  decimals: number;
  aeUsdPrice: string;
}) => {
  if (!reserve || !tokenAePrice || !aeUsdPrice) {
    return '0';
  }

  return new BigNumber(reserve)
    .abs()
    .div(new BigNumber(10).pow(decimals))
    .multipliedBy(tokenAePrice)
    .multipliedBy(aeUsdPrice)
    .toString();
};
