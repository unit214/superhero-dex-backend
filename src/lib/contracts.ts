import NETWORKS from './networks';
import { AeSdk, Node, ContractMethodsBase } from '@aeternity/aepp-sdk';
import { ContractAddress, nonNullable } from './utils';
import * as routerInterface from 'dex-contracts-v2/build/AedexV2Router.aci.json';
import * as factoryInterface from 'dex-contracts-v2/build/AedexV2Factory.aci.json';
import * as pairInterface from 'dex-contracts-v2/build/AedexV2Pair.aci.json';
import ContractWithMethods from '@aeternity/aepp-sdk/es/contract/Contract';

let client: AeSdk;
let node: Node;

export const getClient = async (): Promise<[AeSdk, Node]> => {
  const NETWORK_NAME = nonNullable(process.env.NETWORK_NAME);
  if (!client) {
    node = new Node(NETWORKS[NETWORK_NAME].nodeUrl, {
      ignoreVersion: true,
    });

    client = new AeSdk({
      nodes: [{ name: NETWORK_NAME, instance: node }],
    });
  }
  return [client, node];
};

export type RouterMethods = {
  factory: () => ContractAddress;
};

export type FactoryMethods = {
  get_all_pairs: () => ContractAddress[];
};

export type PairMethods = {
  token0: () => ContractAddress;
  token1: () => ContractAddress;
  total_supply: () => bigint;
  get_reserves: () => {
    reserve0: bigint;
    reserve1: bigint;
  };
};

export type MetaInfo = {
  name: string;
  symbol: string;
  decimals: bigint;
};

export type Aex9Methods = {
  meta_info: () => MetaInfo;
};

export type Context = {
  router: ContractWithMethods<RouterMethods>;
  factory: ContractWithMethods<FactoryMethods>;
  getPair: (
    address: ContractAddress,
  ) => Promise<ContractWithMethods<PairMethods>>;
  getToken: (
    address: ContractAddress,
  ) => Promise<ContractWithMethods<Aex9Methods>>;
  node: Node;
};

const createGetToken =
  (
    tokens: { [key: string]: ContractWithMethods<Aex9Methods> | undefined },
    getInstance: Awaited<ReturnType<typeof instanceFactory>>,
  ) =>
  async (
    tokenAddress: ContractAddress,
  ): Promise<ContractWithMethods<Aex9Methods>> => {
    const cached = tokens[tokenAddress];
    if (cached) {
      return cached;
    }
    const token = await getInstance<Aex9Methods>(pairInterface, tokenAddress);
    tokens[tokenAddress] = token;
    return token;
  };

const createGetPair =
  (
    pairs: { [key: string]: ContractWithMethods<PairMethods> | undefined },
    getInstance: Awaited<ReturnType<typeof instanceFactory>>,
  ) =>
  async (
    pairAddress: ContractAddress,
  ): Promise<ContractWithMethods<PairMethods>> => {
    const cached = pairs[pairAddress];
    if (cached) {
      return cached;
    }
    const pair = await getInstance<PairMethods>(pairInterface, pairAddress);
    pairs[pairAddress] = pair;
    return pair;
  };

const instanceFactory = async (client: AeSdk) => {
  return <T extends ContractMethodsBase>(
    aci: any,
    contractAddress: ContractAddress,
  ) => client.initializeContract<T>({ aci, address: contractAddress });
};

export const getContext = async (): Promise<Context> => {
  const routerAddress = process.env.ROUTER_ADDRESS;
  if (!routerAddress) {
    throw new Error('Router address is not set');
  }
  const [client, node] = await getClient();
  const getInstance = await instanceFactory(client);
  const router = await getInstance<RouterMethods>(
    routerInterface,
    nonNullable<ContractAddress>(routerAddress as ContractAddress),
  );
  const factory = await getInstance<FactoryMethods>(
    factoryInterface,
    nonNullable<ContractAddress>(
      process.env.FACTORY_ADDRESS as ContractAddress,
    ),
  );
  const pairs: { [key: string]: ContractWithMethods<PairMethods> | undefined } =
    {};
  const tokens: {
    [key: string]: ContractWithMethods<Aex9Methods> | undefined;
  } = {};
  return {
    router,
    factory,
    getPair: createGetPair(pairs, getInstance),
    getToken: createGetToken(tokens, getInstance),
    node,
  };
};
