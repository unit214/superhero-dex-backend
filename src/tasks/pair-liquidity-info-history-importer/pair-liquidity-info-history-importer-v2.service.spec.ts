import { Test, TestingModule } from '@nestjs/testing';

import { MdwHttpClientService } from '@/clients/mdw-http-client.service';
import { SdkClientService } from '@/clients/sdk-client.service';
import { PairDbService } from '@/database/pair/pair-db.service';
import { PairLiquidityInfoHistoryV2DbService } from '@/database/pair-liquidity-info-history/pair-liquidity-info-history-v2-db.service';
import { PairLiquidityInfoHistoryV2ErrorDbService } from '@/database/pair-liquidity-info-history-error/pair-liquidity-info-history-v2-error-db.service';
import { bigIntToDecimal } from '@/lib/utils';
import { PairLiquidityInfoHistoryImporterV2Service } from '@/tasks/pair-liquidity-info-history-importer/pair-liquidity-info-history-importer-v2.service';
import resetAllMocks = jest.resetAllMocks;
import {
  contractLog1,
  contractLog2,
  contractLog3,
  contractLog4,
  contractLog5,
  contractLog6,
  contractLog7,
  contractLog8,
  initialMicroBlock,
  pairContract,
  pairWithTokens,
} from '@/test/mock-data/pair-liquidity-info-history-mock-data';

const mockMdwClient = {
  getContract: jest.fn(),
  getMicroBlock: jest.fn(),
  getContractLogsUntilCondition: jest.fn(),
};

const mockPairDb = { getAll: jest.fn() };

const mockPairLiquidityInfoHistoryV2Db = {
  getLastlySyncedLogByPairId: jest.fn(),
  upsert: jest.fn(),
};

const mockPairLiquidityInfoHistoryV2ErrorDb = {
  getErrorWithinHours: jest.fn(),
  upsert: jest.fn(),
};

describe('PairLiquidityInfoHistoryImporterV2Service', () => {
  let service: PairLiquidityInfoHistoryImporterV2Service;
  let logSpy: jest.SpyInstance;
  let errorSpy: jest.SpyInstance;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [
        PairLiquidityInfoHistoryImporterV2Service,
        { provide: MdwHttpClientService, useValue: mockMdwClient },
        { provide: PairDbService, useValue: mockPairDb },
        {
          provide: PairLiquidityInfoHistoryV2DbService,
          useValue: mockPairLiquidityInfoHistoryV2Db,
        },
        {
          provide: PairLiquidityInfoHistoryV2ErrorDbService,
          useValue: mockPairLiquidityInfoHistoryV2ErrorDb,
        },
        SdkClientService,
      ],
    }).compile();
    service = module.get<PairLiquidityInfoHistoryImporterV2Service>(
      PairLiquidityInfoHistoryImporterV2Service,
    );
    logSpy = jest.spyOn(service.logger, 'log');
    errorSpy = jest.spyOn(service.logger, 'error');
    resetAllMocks();
  });

  describe('import', () => {
    it('should import liquidity correctly', async () => {
      // Mock functions
      mockPairDb.getAll.mockResolvedValue([pairWithTokens]);
      mockPairLiquidityInfoHistoryV2ErrorDb.getErrorWithinHours.mockResolvedValue(
        undefined,
      );
      mockPairLiquidityInfoHistoryV2Db.getLastlySyncedLogByPairId
        .mockResolvedValueOnce(undefined)
        .mockResolvedValueOnce({
          reserve0: bigIntToDecimal(100n),
          reserve1: bigIntToDecimal(100n),
        });
      mockMdwClient.getContract.mockResolvedValue(pairContract);
      mockMdwClient.getMicroBlock.mockResolvedValue(initialMicroBlock);
      mockPairLiquidityInfoHistoryV2Db.upsert.mockResolvedValue(null);
      mockMdwClient.getContractLogsUntilCondition.mockResolvedValue([
        contractLog1,
        contractLog2,
        contractLog3,
        contractLog4,
        contractLog5,
        contractLog6,
        contractLog7,
        contractLog8,
      ]);

      // Start import
      await service.import();

      // Assertions
      expect(
        mockPairLiquidityInfoHistoryV2ErrorDb.getErrorWithinHours,
      ).toHaveBeenCalledTimes(5); // Once for pair and 4 times for each inserted event

      expect(logSpy.mock.calls).toEqual([
        ['Started syncing pair liquidity info history.'],
        ['Syncing liquidity info history for 1 pairs.'],
        [
          `Inserted initial liquidity for pair ${pairWithTokens.id} ${pairWithTokens.address}.`,
        ],
        [
          `Completed sync for pair ${pairWithTokens.id} ${pairWithTokens.address}. Synced 4 log(s).`,
        ],
        ['Finished liquidity info history sync for all pairs.'],
      ]);

      expect(mockPairLiquidityInfoHistoryV2Db.upsert).toHaveBeenCalledTimes(5);
      expect(
        mockPairLiquidityInfoHistoryV2Db.upsert.mock.calls,
      ).toMatchSnapshot();
    });

    it('should skip a pair if there was a recent error', async () => {
      // Mock functions
      mockPairDb.getAll.mockResolvedValue([pairWithTokens]);
      mockPairLiquidityInfoHistoryV2ErrorDb.getErrorWithinHours.mockResolvedValue(
        {
          id: 1,
          pairId: 1,
          microBlockHash: '',
          transactionHash: '',
          logIndex: -1,
          error: 'error',
          timesOccurred: 1,
          createdAt: Date.now(),
          updatedAt: Date.now(),
        },
      );

      // Start import
      await service.import();

      // Assertions
      expect(logSpy.mock.calls).toEqual([
        ['Started syncing pair liquidity info history.'],
        ['Syncing liquidity info history for 1 pairs.'],
        [`Skipped pair ${pairWithTokens.id} due to recent error.`],
        ['Finished liquidity info history sync for all pairs.'],
      ]);
    });

    it('should skip a log if there was a recent error', async () => {
      // Mock functions
      mockPairDb.getAll.mockResolvedValue([pairWithTokens]);
      mockPairLiquidityInfoHistoryV2ErrorDb.getErrorWithinHours
        .mockResolvedValueOnce(undefined)
        .mockResolvedValueOnce({
          id: 1,
          pairId: 1,
          microBlockHash: contractLog1.block_hash,
          transactionHash: contractLog1.call_tx_hash,
          logIndex: contractLog1.log_idx,
          error: 'error',
          timesOccurred: 1,
          createdAt: Date.now(),
          updatedAt: Date.now(),
        })
        .mockResolvedValueOnce(undefined);
      mockPairLiquidityInfoHistoryV2Db.getLastlySyncedLogByPairId.mockResolvedValue(
        {},
      );
      mockPairLiquidityInfoHistoryV2Db.upsert.mockResolvedValue(null);
      mockMdwClient.getContractLogsUntilCondition.mockResolvedValue([
        contractLog1,
        contractLog2,
        contractLog4,
        contractLog5,
      ]);

      // Start import
      await service.import();

      // Assertions
      expect(
        mockPairLiquidityInfoHistoryV2ErrorDb.getErrorWithinHours,
      ).toHaveBeenCalledTimes(3); // Once for pair and 2 times for each event

      expect(logSpy.mock.calls).toEqual([
        ['Started syncing pair liquidity info history.'],
        ['Syncing liquidity info history for 1 pairs.'],
        [
          `Skipped log with block hash ${contractLog1.block_hash} tx hash ${contractLog1.call_tx_hash} and log index ${contractLog1.log_idx} due to recent error.`,
        ],
        [
          `Completed sync for pair ${pairWithTokens.id} ${pairWithTokens.address}. Synced 1 log(s).`,
        ],
        ['Finished liquidity info history sync for all pairs.'],
      ]);

      expect(
        mockPairLiquidityInfoHistoryV2Db.upsert.mock.calls,
      ).toMatchSnapshot();
    });

    it('should catch and insert an error on pair level', async () => {
      const error = {
        pairId: pairWithTokens.id,
        microBlockHash: '',
        transactionHash: '',
        logIndex: -1,
        error: 'Error: error',
      };

      // Mock functions
      mockPairDb.getAll.mockResolvedValue([pairWithTokens]);
      mockPairLiquidityInfoHistoryV2ErrorDb.getErrorWithinHours.mockResolvedValue(
        undefined,
      );
      mockPairLiquidityInfoHistoryV2Db.getLastlySyncedLogByPairId.mockResolvedValue(
        {},
      );
      mockMdwClient.getContractLogsUntilCondition.mockRejectedValue(
        new Error('error'),
      );
      mockPairLiquidityInfoHistoryV2ErrorDb.upsert.mockResolvedValue(null);

      // Start import
      await service.import();

      // Assertions
      expect(mockPairLiquidityInfoHistoryV2ErrorDb.upsert).toHaveBeenCalledWith(
        error,
      );

      expect(logSpy.mock.calls).toEqual([
        ['Started syncing pair liquidity info history.'],
        ['Syncing liquidity info history for 1 pairs.'],
        ['Finished liquidity info history sync for all pairs.'],
      ]);

      expect(errorSpy.mock.calls).toEqual([
        [`Skipped pair. ${JSON.stringify(error)}`],
      ]);
    });

    it('should catch and insert an error on log level', async () => {
      const error = {
        pairId: pairWithTokens.id,
        microBlockHash: contractLog3.block_hash,
        transactionHash: contractLog3.call_tx_hash,
        logIndex: parseInt(contractLog3.log_idx),
        error: 'Error: error',
      };

      // Mock functions
      mockPairDb.getAll.mockResolvedValue([pairWithTokens]);
      mockPairLiquidityInfoHistoryV2ErrorDb.getErrorWithinHours.mockResolvedValue(
        undefined,
      );
      mockPairLiquidityInfoHistoryV2Db.getLastlySyncedLogByPairId
        .mockResolvedValueOnce({})
        .mockRejectedValueOnce(new Error('error'))
        .mockRejectedValueOnce(undefined);
      mockMdwClient.getContractLogsUntilCondition.mockResolvedValue([
        contractLog3,
        contractLog4,
        contractLog5,
      ]);
      mockPairLiquidityInfoHistoryV2Db.upsert.mockResolvedValue(null);
      mockPairLiquidityInfoHistoryV2ErrorDb.upsert.mockResolvedValue(null);

      // Start import
      await service.import();

      // Assertions
      expect(mockPairLiquidityInfoHistoryV2ErrorDb.upsert).toHaveBeenCalledWith(
        error,
      );

      expect(logSpy.mock.calls).toEqual([
        ['Started syncing pair liquidity info history.'],
        ['Syncing liquidity info history for 1 pairs.'],
        [
          `Completed sync for pair ${pairWithTokens.id} ${pairWithTokens.address}. Synced 1 log(s).`,
        ],
        ['Finished liquidity info history sync for all pairs.'],
      ]);

      expect(errorSpy.mock.calls).toEqual([
        [`Skipped log. ${JSON.stringify(error)}`],
      ]);

      expect(
        mockPairLiquidityInfoHistoryV2Db.upsert.mock.calls,
      ).toMatchSnapshot();
    });
  });
});
