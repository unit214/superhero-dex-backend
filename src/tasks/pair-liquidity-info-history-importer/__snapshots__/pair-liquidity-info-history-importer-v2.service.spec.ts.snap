// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PairLiquidityInfoHistoryImporterV2Service import should import liquidity correctly 1`] = `
[
  [
    {
      "aeUsdPrice": "0.050559",
      "deltaReserve0": "0",
      "deltaReserve1": "0",
      "eventType": "CreatePair",
      "height": 10000,
      "logIndex": 0,
      "microBlockHash": "mh_hash0",
      "microBlockTime": 1000000000000n,
      "pairId": 1,
      "reserve0": "0",
      "reserve1": "0",
      "transactionHash": "th_",
      "transactionIndex": 0n,
    },
  ],
  [
    {
      "aeUsdPrice": "0.050559",
      "deltaReserve0": "100",
      "deltaReserve1": "100",
      "eventType": "PairMint",
      "height": 10001,
      "logIndex": 2,
      "microBlockHash": "mh_hash1",
      "microBlockTime": 1000000000001n,
      "pairId": 1,
      "reserve0": "100",
      "reserve1": "100",
      "transactionHash": "th_hash1",
      "transactionIndex": 10000001n,
    },
  ],
  [
    {
      "aeUsdPrice": "0.050559",
      "deltaReserve0": "100",
      "deltaReserve1": "100",
      "eventType": "Sync",
      "height": 20002,
      "logIndex": 1,
      "microBlockHash": "mh_hash2",
      "microBlockTime": 2000000000002n,
      "pairId": 1,
      "reserve0": "200",
      "reserve1": "200",
      "transactionHash": "th_hash2",
      "transactionIndex": 20000002n,
    },
  ],
  [
    {
      "aeUsdPrice": "0.050559",
      "deltaReserve0": "1",
      "deltaReserve1": "-1",
      "eventType": "SwapTokens",
      "height": 30003,
      "logIndex": 2,
      "microBlockHash": "mh_hash3",
      "microBlockTime": 3000000000003n,
      "pairId": 1,
      "reserve0": "201",
      "reserve1": "199",
      "transactionHash": "th_hash3",
      "transactionIndex": 30000003n,
    },
  ],
  [
    {
      "aeUsdPrice": "0.050559",
      "deltaReserve0": "-101",
      "deltaReserve1": "-99",
      "eventType": "PairBurn",
      "height": 30003,
      "logIndex": 2,
      "microBlockHash": "mh_hash3",
      "microBlockTime": 3000000000003n,
      "pairId": 1,
      "reserve0": "100",
      "reserve1": "100",
      "transactionHash": "th_hash4",
      "transactionIndex": 40000004n,
    },
  ],
]
`;

exports[`PairLiquidityInfoHistoryImporterV2Service import should skip a log if there was a recent error 1`] = `
[
  [
    {
      "aeUsdPrice": "0.050559",
      "deltaReserve0": "1",
      "deltaReserve1": "-1",
      "eventType": "SwapTokens",
      "height": 30003,
      "logIndex": 2,
      "microBlockHash": "mh_hash3",
      "microBlockTime": 3000000000003n,
      "pairId": 1,
      "reserve0": "201",
      "reserve1": "199",
      "transactionHash": "th_hash3",
      "transactionIndex": 30000003n,
    },
  ],
]
`;

exports[`PairLiquidityInfoHistoryImporterV2Service import should catch and insert an error on log level 1`] = `
[
  [
    {
      "aeUsdPrice": "0.050559",
      "deltaReserve0": "1",
      "deltaReserve1": "-1",
      "eventType": "SwapTokens",
      "height": 30003,
      "logIndex": 2,
      "microBlockHash": "mh_hash3",
      "microBlockTime": 3000000000003n,
      "pairId": 1,
      "reserve0": "201",
      "reserve1": "199",
      "transactionHash": "th_hash3",
      "transactionIndex": 30000003n,
    },
  ],
]
`;
